visualizer/actions.js

```diff
@@ -100,7 +114,85 @@ function createActions ({flamegraph, state}, emit) {
         state.typeFilters.exclude.add(name)
       }
       if (state.control.tiers) state.typeFilters.bgs = highlightTypeFilters()
+      save()
       emit(state)
     }
   }
+
+  function pushState () {
+    return () => {
+      const { merged } = state.control
+      const excludeTypes = Array.from(state.typeFilters.exclude)
+      const { nodesToIds } = merged ? mergedTags : unmergedTags
+      const historyState = {
+        merged,
+        nodeId: nodesToIds.get(state.focusedNodeId),
```

Comment:
There appears to be some inconsistency about `state.focusedNodeId`'s type: 

- This code seems to assume it's a node object (`tagNodesWithIds` sets `nodesToIds` keys as tree layers), and it appears that `flamegraph.on('zoom', (d) => { ... focusNode(d) }` in index.js:29-36 sets it as a node object 
 - But the hash JSON parser and `window.addEventListener('popstate'` listener in index.js set it as a numeric ID, and the property name suggests it will be a number

The code seems to work fine despite this - it seems that the `pushState` route consistently expects and gets `state.focusedNodeId` as a node object and the `jumpToState` route consistently expects and gets it as a number, and the two don't currently collide - but it seems like a pit trap to have a property that appears to be a numeric id, when by some routes it could actually be a node object?

Reply:
Actually I think it is possible for the two routes to collide. To replicate:

- Click on some node. `state.focusedNodeId` is set as an object via the `pushState` route
- Toggle a visibility setting. Keeps the ID
- Undo the visibility setting with the back button. `state.focusedNodeId` is set as a number via the `jumpToState ` route
- Toggle another visibility setting. This unexpectedly erases the node Id from the hash - presumably, because it's going down the `pushState` route and doing `nodesToIds.get()` on a numeric id, which returns undefined?

Reply:
good find, in my head `focusedNodeId` was always a number but clearly that's not what I actually wrote down :sweat_smile: thanks!